src/widgets/maybe.rs

```diff
@@ -0,0 +1,167 @@
+
+/// A widget that switches between two possible child views, for `Data` that
+/// is `Option<T>`.
+pub struct Maybe<T> {
+    some_maker: Box<dyn Fn() -> Box<dyn Widget<T>>>,
+    none_maker: Box<dyn Fn() -> Box<dyn Widget<()>>>,
+    widget: MaybeWidget<T>,
+}
+
+enum MaybeWidget<T> {
+    Some(WidgetPod<T, Box<dyn Widget<T>>>),
+    None(WidgetPod<(), Box<dyn Widget<()>>>),
+}
+
+impl<T: Data> Maybe<T> {
+    /// Create a new `Maybe` widget with a `Some` and a `None` branch.
+    pub fn new<W1, W2>(
+        // we make these generic so that the caller doesn't have to explicitly
+        // box. We don't technically *need* to box, but it seems simpler.
+        some_maker: impl Fn() -> W1 + 'static,
+        none_maker: impl Fn() -> W2 + 'static,
+    ) -> Maybe<T>
+    where
+        W1: Widget<T> + 'static,
+        W2: Widget<()> + 'static,
+    {
+        let widget = MaybeWidget::Some(WidgetPod::new(some_maker().boxed()));
+        Maybe {
+            some_maker: Box::new(move || some_maker().boxed()),
+            none_maker: Box::new(move || none_maker().boxed()),
+            widget,
+        }
+    }
+
+    /// Create a new `Maybe` widget where the `None` branch is an empty widget.
+    #[allow(dead_code)]
+    pub fn or_empty<W1: Widget<T> + 'static>(some_maker: impl Fn() -> W1 + 'static) -> Maybe<T> {
+        Self::new(some_maker, || SizedBox::empty())
+    }
+
+    fn rebuild_widget(&mut self, is_some: bool) {
+        if is_some {
+            self.widget = MaybeWidget::Some(WidgetPod::new((self.some_maker)()));
+        } else {
+            self.widget = MaybeWidget::None(WidgetPod::new((self.none_maker)()));
+        }
+    }
+}
+
+impl<T: Data> Widget<Option<T>> for Maybe<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut Option<T>, env: &Env) {
+        match data.as_mut() {
+            Some(d) => self.widget.unwrap_some().event(ctx, event, d, env),
+            None => self.widget.unwrap_none().event(ctx, event, &mut (), env),
+        }
+    }
+
+    fn lifecycle(
+        &mut self,
+        ctx: &mut LifeCycleCtx,
+        event: &LifeCycle,
+        data: &Option<T>,
+        env: &Env,
+    ) {
+        if let LifeCycle::WidgetAdded = event {
+            if data.is_some() != self.widget.is_some() {
+                // only possible at launch, because we default to `Some`
+                self.rebuild_widget(data.is_some());
+            }
+        }
+        match data.as_ref() {
+            Some(d) => self.widget.unwrap_some().lifecycle(ctx, event, d, env),
+            None => self
+                .widget
+                .unwrap_none()
+                .lifecycle(ctx, event, &mut (), env),
+        }
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &Option<T>, data: &Option<T>, env: &Env) {
+        if old_data.is_some() != data.is_some() {
+            self.rebuild_widget(data.is_some());
+            ctx.children_changed();
+            ctx.request_layout();
```

Comment:
Does `children_changed` not include the effect of `request_layout`? I would expect `request_layout` to happen whenever `children_changed` is called.

Reply:
good catch, yes, I believe we always invalidate everything when children changed. I think this happens automatically, by having `needs_layout` be `true` for a new `WidgetPod`. it isn't set otherwise, though, and that might be a bug? Or certainly cause confusion in some circumstances.